### PR TITLE
JAX-WS: Prevent CXF from ignoring imported WDSL Namespaces used by <wsdl:header> message parts

### PR DESCRIPTION
**Pull request by neuwerk**
CXF will ignore the Out Of Bounds namespaces for Header parts that are imported from another WSDL. This pull request forces CXF to check the Out of Bounds MessageParts to correct the QNames and allow a header parts to be properly marshalled and unmashalled by CXF. 

This fixes ClassCastExceptions when trying obtain any Header type from a javax.xml.ws.Holder object, an example is below. 
`[err] java.lang.ClassCastException: java.lang.Object incompatible with com.example.Header`

### CODE REVIEW

**Comment by neuwerk** [comment]
#build
#spawn.fullfat.buckets=com.ibm.ws.jaxws.2.2.webcontainer_fat, com.ibm.ws.jaxws.2.2.webcontainer_fat_extended, com.ibm.ws.jaxws.clientcontainer_fat, com.ibm.ws.jaxws.cdi_fat, com.ibm.ws.jaxws.ejb_fat, com.ibm.ws.jaxws.X.wsat_fat, com.ibm.ws.jaxrs.2.0_fat, com.ibm.ws.jaxrs.2.1_fat, com.ibm.ws.jaxrs.2.1_fat_executor, io.openliberty.ws.jaxrs.2.1.jaxws.2.2_fat, io.openliberty.ws.jaxrs.global.handler.internal_fat, io.openliberty.ws.jaxws.global.handler.internal_fat, com.ibm.ws.wsat_fat.db,com.ibm.ws.wssecurity_fat.wsscxf.saml.3,com.ibm.ws.wssecurity_fat.wsscxf.1,com.ibm.ws.wssecurity_fat.wsscxf.saml.2,com.ibm.ws.wssecurity_fat.wsscxf.saml.4,com.ibm.ws.wssecurity_fat.wsscxf,com.ibm.ws.wssecurity_fat.wsscxf.saml.1,com.ibm.ws.wssecurity_fat.wsscxf.saml,com.ibm.ws.wssecurity_fat.wsscxf.saml.5, com.ibm.ws.jaxrs.2.0_fat,com.ibm.ws.jaxrs.2.0.cdi.1.2_fat,com.ibm.ws.jaxrs.2.0.client_fat,com.ibm.ws.jaxrs.2.1_fat,com.ibm.ws.jaxrs.2.1_fat_executor,com.ibm.ws.jaxrs.2.1_fat_extended,com.ibm.ws.jaxrs.2.1.cdi.2.0_fat,com.ibm.ws.jaxrs.2.1.client_fat,com.ibm.ws.jaxrs.2.1.sse_fat,com.ibm.ws.jaxrs.2.x_fat_client, io.openliberty.jaxws.security_fat, io.openliberty.jaxws.security_fat.1, io.openliberty.jaxws.security_fat.ssl

File: dev/com.ibm.ws.org.apache.cxf.cxf.rt.bindings.soap.3.2/src/org/apache/cxf/binding/soap/SoapBindingFactory.java [hunk 1]
```diff
@@ -745,13 +763,25 @@ private void initializeMessage(SoapBindingInfo bi, BindingOperationInfo boi, Bin
                         if (mpi.getName().getLocalPart().equals(header.getPart())
                             && mpi.getMessageInfo().getName().equals(header.getMessage())) {
                             part = mpi;
+                            isOutofBoundPart = true;
+                            outOfBoundQname = mpi.getName();
                         }
                     }
                 }
                 if (part != null) {
```
Comment (sawadood) [inline dev/com.ibm.ws.org.apache.cxf.cxf.rt.bindings.soap.3.2/src/org/apache/cxf/binding/soap/SoapBindingFactory.java:771]: In initializeMessage(), duplicate message parts are added to the soap header. This code should be changed as follows:

      if (part != null) {
            if (isOutofBoundPart) {
                  LOG.finest("This Header's QName is mismatched from the expected OutOfBounds QName, removing mismatched MessagePartInfo, and adding mismatched Header");
                  part = msg.getMessagePart(new QName(msg.getName().getNamespaceURI(), header.getPart()));                                        
                  headerInfo.setPart(part);
                  messageParts.remove(part);
                  bmsg.addExtensor(headerInfo);
            }
            else {
                  headerInfo.setPart(part);
                  messageParts.remove(part);
                  bmsg.addExtensor(headerInfo);
            }   // Liberty Change End
      }
  
      With this change, update to SoapHeaderInterceptor.java will not be required.

File: dev/com.ibm.ws.org.apache.cxf.cxf.rt.bindings.soap.3.2/src/org/apache/cxf/binding/soap/interceptor/SoapHeaderInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,232 @@
+/**
```
Comment (sawadood) [inline dev/com.ibm.ws.org.apache.cxf.cxf.rt.bindings.soap.3.2/src/org/apache/cxf/binding/soap/interceptor/SoapHeaderInterceptor.java]: This update will not be required if the update I mentioned in SoapBindingFactory.java is made.

**Comment by neuwerk** [comment]
#build
#spawn.fullfat.buckets=com.ibm.ws.jaxws.2.2.webcontainer_fat, com.ibm.ws.jaxws.2.2.webcontainer_fat_extended, com.ibm.ws.jaxws.clientcontainer_fat, com.ibm.ws.jaxws.cdi_fat, com.ibm.ws.jaxws.ejb_fat, com.ibm.ws.jaxws.X.wsat_fat, com.ibm.ws.jaxrs.2.0_fat, com.ibm.ws.jaxrs.2.1_fat, com.ibm.ws.jaxrs.2.1_fat_executor, io.openliberty.ws.jaxrs.2.1.jaxws.2.2_fat, io.openliberty.ws.jaxrs.global.handler.internal_fat, io.openliberty.ws.jaxws.global.handler.internal_fat, com.ibm.ws.wsat_fat.db,com.ibm.ws.wssecurity_fat.wsscxf.saml.3,com.ibm.ws.wssecurity_fat.wsscxf.1,com.ibm.ws.wssecurity_fat.wsscxf.saml.2,com.ibm.ws.wssecurity_fat.wsscxf.saml.4,com.ibm.ws.wssecurity_fat.wsscxf,com.ibm.ws.wssecurity_fat.wsscxf.saml.1,com.ibm.ws.wssecurity_fat.wsscxf.saml,com.ibm.ws.wssecurity_fat.wsscxf.saml.5, com.ibm.ws.jaxrs.2.0_fat,com.ibm.ws.jaxrs.2.0.cdi.1.2_fat,com.ibm.ws.jaxrs.2.0.client_fat,com.ibm.ws.jaxrs.2.1_fat,com.ibm.ws.jaxrs.2.1_fat_executor,com.ibm.ws.jaxrs.2.1_fat_extended,com.ibm.ws.jaxrs.2.1.cdi.2.0_fat,com.ibm.ws.jaxrs.2.1.client_fat,com.ibm.ws.jaxrs.2.1.sse_fat,com.ibm.ws.jaxrs.2.x_fat_client, io.openliberty.jaxws.security_fat, io.openliberty.jaxws.security_fat.1, io.openliberty.jaxws.security_fat.ssl

**Comment by neuwerk** [comment]
Failures are unrelated, I just had fat buckets named that no longer exist in open liberty.

**Review by sawadood — APPROVED** [review]
Changes look good

**Comment by neuwerk** [comment]
#libby

**Pull request merged**